alternator: implement TableClass attribute by changing compression

nyh opened the pull request:
DynamoDB announced in December 2021 a new attribute of a table, called table class. The choice of "table class" doesn't change anything in the table's capabilities - it just controls the pricing of storage and access: There are currently two options - STANDARD (the default) and STANDARD_INFREQUENT_ACCESS. The latter is for infrequently-accessed tables, its the data storage costs are lower (in DynamoDB, by 60%), while reads and writes are more expensive (in DynamoDB, by 25%).

In this pull request we map the TableClass option in Scylla to different sstable compression settings:

STANDARD uses to the existing default compression of lz4-with-dictionaries, which gives reasonable compression ratios at small CPU overhead.

STANDARD_INFREQUENT_ACCESS uses zstd-with-dictionaries compression, which gives better compression ratios, with higher overhead.

There are a couple of insignificant deviations from DynamoDB in our implementation (which are described in the documentation in the last patch, as well as in xfailing tests in the first patch), but a more significant deviation is the following:

In DynamoDB, changing the table class via `UpdateTable` triggers a background  operation to change the table's storage, and is limited to two changes per 30-day period. In Alternator, the compression setting in the schema is updated immediately, but existing SSTables are **not** rewritten; only new SSTables written by subsequent writes and compactions will use the new compression.

Fixes #10431

nyh committed:
alternator: tests for new TableClass attribute

DynamoDB announced in December 2021 a new attribute of a table, called
table class. The choice of "table class" doesn't change anything in the
table's capabilities - it just controls the pricing of storage and access:
There are currently two options - STANDARD (the default) and
STANDARD_INFREQUENT_ACCESS. The latter is for infrequently-accessed tables,
its the data storage costs are lower (in DynamoDB, by 60%), while reads and
writes are more expensive (in DynamoDB, by 25%).

We plan in the next patch to map the TableClass option in Scylla to
different sstable compression settings: STANDARD will use the existing
default compression of lz4, which gives reasonable compression ratios
at small CPU overhead, while STANDARD_INFREQUENT_ACCESS will use zstd
compressions, which gives better compression ratios, with higher CPU
overhead.

So in this patch we introduce tests for the table class feature.
Most tests check how this feature works on DynamoDB, so now (before
we implemented this feature in Alternator) xfail on Alternator, and
pass on DynamoDB. The last two tests check how the table class affects
the sstable compression in ScyllaDB, so they are scylla_only tests
(and xfail until we actually implement this feature in the next patch).

Refs #10431

Signed-off-by: Nadav Har'El <nyh@scylladb.com>

nyh committed:
alternator: implement TableClass attribute by changing compression

DynamoDB announced in December 2021 a new attribute of a table, called
table class. The choice of "table class" doesn't change anything in the
table's capabilities - it just controls the pricing of storage and
access: There are currently two options - STANDARD (the default) and
STANDARD_INFREQUENT_ACCESS. The latter is for infrequently-accessed
tables, its the data storage costs are lower (in DynamoDB, by 60%),
while reads and writes are more expensive (in DynamoDB, by 25%).

In this patch we implement the TableClass option in CreateTable, mapping
the TableClass option in Scylla to different sstable compression settings:

* STANDARD uses to the existing default compression of lz4 with
  dictionaries, which gives reasonable compression ratios at small
  CPU overhead.

* STANDARD_INFREQUENT_ACCESS uses zstd compressions with dictionaries,
  which gives better compression ratios, with higher CPU and memory
  overhead.

Fixes #10431

nyh committed:
alternator: implement TableClass in UpdateTable

In the previous patch we supported the TableClass setting in CreateTable
to determine whether to use standard compression (STANDARD) or exta
compression (STANDARD_INFREQUENT_ACCESS). In this patch we also implement
this in UpdateTable - allow a user to change the compression on an
already-existing table.

This TableClass is a DynamoDB feature that Alternator interprets to
involve compression.

Note in DynamoDB, when the TableClass is modified with UpdateTable,
DynamoDB starts a background operation to restructure the existing
table (this is a heavy operation, so is only allowed twice per day).
This is important because DynamoDB changes for STANDARD_INFREQUENT_ACCESS
lower storage costs, so whatever it does to make storage cheaper
(Alternator uses extra compression), it must do this for the entire
existing table. In contrast, currently in Alternator the update does
NOT modify the existing tables, and only affects new sstables written
during further writes and compactions.

Another small difference in our implementation is that DynamoDB returns
in DescribeTable's TableClassSummary a field LastUpdateDateTime saying
when the TableClass of this table was last updated. We haven't
implemented this field.

nyh committed:
docs/alternator: document the TableClass feature

docs/alternator/compatibility.md said that DynamoDB's feature
of "TableClass" was not implemented. Now that we have implemented it,
replace that statement by a section describing this feature, and in
what ways our implementation differs from DynamoDB's.

nyh commented on alternator/executor.cc:
```diff
@@ -1448,6 +1480,17 @@ future<executor::request_return_type> executor::create_table_on_shard0(service::
+        case table_class::standard:
+            builder.set_compressor_params(compression_parameters::algorithm::lz4_with_dicts);
+            break;
+        }
```
Good catch, although it's a bit sad we need to care of such low-level details in Alternator. I'll fix.

nyh commented on alternator/executor.cc:
```diff
@@ -517,6 +517,13 @@ future<rjson::value> executor::fill_table_description(schema_ptr schema, table_s
     } else {
         rjson::add(table_description["BillingModeSummary"], "BillingMode", "PROVISIONED");
     }
+    auto compression_algorithm = schema->get_compressor_params().get_algorithm();
+    if (compression_algorithm == compression_parameters::algorithm::zstd_with_dicts ||
+            compression_algorithm == compression_parameters::algorithm::zstd) {
+        rjson::value table_class_summary = rjson::empty_object();
+        rjson::add(table_class_summary, "TableClass", "STANDARD_INFREQUENT_ACCESS");
+        rjson::add(table_description, "TableClassSummary", std::move(table_class_summary));
+    }
```
I really wanted to avoid having yet another tag for this purpose.
Anyway, if some option option like `sstable_compression_user_table_options` (which I didn't know exists) modifies the default compression, it sounds like it will really change the default table creation from STANDARD to STANDARD_INFREQUENT_ACCESS. I'm not sure it's a bug, it's more like a feature?
Sure, it will be different from DynamoDB's default, but it's the user who changed this default - apparently for a reason!
In any case, it reminds me I should test this case. I'll add a test.

nyh replied:
Turns out it's not possible to efficiently write this such a test because sstable_compression_user_table_options is not live-updatable (see https://github.com/scylladb/scylladb/pull/26003). I could write a test/cluster test just for this feature, but I don't think it's important enough.

nyh commented on test/alternator/test_tableclass.py:
```diff
@@ -0,0 +1,233 @@
+# to finish. While the update is happening, the table's TableStatus is
+# changed from ACTIVE to UPDATING - and we need to wait for it to become
+# ACTIVE again.
+def wait_for_active(table):
+    timeout = time.time() + 60
+    while time.time() < timeout:
+        desc = table.meta.client.describe_table(TableName=table.name)
+        if desc['Table']['TableStatus'] == 'ACTIVE':
+            return
+        time.sleep(1)
+    raise AssertionError('wait_for_active did not complete until timeout')
```
In practice, 60 seconds worked just fine for me. I'd rather not prematurely fix a problem we don't encounter.

nyh commented on test/alternator/test_tableclass.py:
```diff
@@ -0,0 +1,233 @@
+        'AttributeDefinitions': [{ 'AttributeName': 'p', 'AttributeType': 'S'}],
+        'TableClass' : 'STANDARD'
+    }
+    with new_test_table(dynamodb, **schema) as table:
+        assert 'LZ4WithDictsCompressor' in scylla_get_compression(dynamodb, table)
```
This can't happen on a modern cluster. No need to handle this in the test.

annastuchlik reviewed (commented):
The documentation looks good. 
The only thing I have concerns above is: "In DynamoDB, the latter reduces storage costs by
60% in exchange for 25% higher per-request costs."

Do we have any clear evidence for that? 
Does it apply to any use case (is it always true)?
Is there any value in providing information about DynamoDB? 

I think we should avoid embedding DynamoDB-specific pricing numbers that could change over time. We can describe the tradeoff without providing any numbers:
**_In DynamoDB, the latter reduces storage costs in exchange for higher request costs._**
Or we could remove the sentence altogether.

@mykaul What do you think?

radoslawcybulski commented on alternator/executor.cc:
```diff
@@ -2252,7 +2294,7 @@ future<executor::request_return_type> executor::update_table(client_state& clien
 
             if (empty_request) {
-                co_return api_error::validation("UpdateTable requires one of GlobalSecondaryIndexUpdates, VectorIndexUpdates, StreamSpecification or BillingMode to be specified");
+                co_return api_error::validation("UpdateTable requires one of GlobalSecondaryIndexUpdates, VectorIndexUpdates, StreamSpecification, BillingMode or TableClass to be specified");
```
Nit: `at least one`

nyh replied:
Interesting. I was sure that it's exactly one of, and that we even have tests for this, but now I realize that actually we don't. I'll look into this and add tests whether it's allowed to change TableClass (for example) together with another change.

nyh replied:
Turns out you **can't** combine TableClass modification with any other - you get:
**TableClass modification must be the only operation in the request**.

I still need to check this for the other types, and also fix the code to check this and pass the new test. But this error message is indeed correct. Or mostly correct.

nyh replied:
Add a new fifth patch with tests that verify that in fact (despite the lack of DynamoDB documentation on this topic) various combinations of two UpdateTable operations are not allowed, and also fixing some places where we forgot checking these forbidden combinations.

radoslawcybulski replied:
In that case could we get `exactly one` instead?

nyh replied:
But when the message said "one" you already thought it meant "exactly one" (and therefore suggested I changed it to "at least one")...

radoslawcybulski replied:
True i guess. `exactly one` removes all doubts (in other words `exactly` makes `one` not accidental), that's why i'm suggesting. Not a strong suggestion tho.

nyh committed:
alternator: UpdateTable should only allow one operation

If an UpdateTable request is empty, we print the error message:

    UpdateTable requires one of GlobalSecondaryIndexUpdates,
    VectorIndexUpdates, StreamSpecification, BillingMode or TableClass
    to be specified

A reviewer asked if exactly one operation is allowed, or maybe it should
be "at least one".

Although it isn't documented anywhere in DynamoDB's documentation, the
answer is that only a single operation is allowed - you can't combine
two different operations like TableClass and GlobalSecondaryIndexUpdates
in one operation.

This patch adds to our code some missing checks for some forbidden
combinations, and also adds tests (which, as usual, pass on DynamoDB
and after this patch also on Alternator) for these forbidden combinations.

Signed-off-by: Nadav Har'El <nyh@scylladb.com>

nyh commented on test/alternator/test_tableclass.py:
```diff
@@ -0,0 +1,250 @@
+# Copyright 2024-present ScyllaDB
```
As you can see in the pull request, unfortunately this is genuinely a pull request that started two years ago.

annastuchlik reviewed (approved):
If the numbers are reliable (and you've convinced me they are), I'm OK with leaving them (until someone objects).
Thanks!
